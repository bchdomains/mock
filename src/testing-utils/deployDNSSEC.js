const ROOT_NODE = '0x00000000000000000000000000000000'
async function deployDNSSEC(web3, accounts, ens) {
  console.log('***DnsRegistrar', {DnsRegistrar})
  console.log('***DNSSEC', {DNSSEC})

  const { sha3 } = web3.utils
  function deploy(contractJSON, ...args) {
    const contract = new web3.eth.Contract(contractJSON.abi)
    return contract
      .deploy({
        data: contractJSON.bytecode,
        arguments: args
      })
      .send({
        from: accounts[0],
        gas: 6700000
      })
  }

  function namehash(name) {
    let node =
      '0x0000000000000000000000000000000000000000000000000000000000000000'
    if (name !== '') {
      let labels = name.split('.')
      for (let i = labels.length - 1; i >= 0; i--) {
        node = sha3(node + sha3(labels[i]).slice(2), {
          encoding: 'hex'
        })
      }
    }
    return node.toString()
  }

  function loadContract(modName, contractName) {
    return require(`@ensdomains/${modName}/build/contracts/${contractName}`)
  }

  const RSASHA256Algorithm = loadContract('dnssec-oracle', 'RSASHA256Algorithm')
  const RSASHA1Algorithm = loadContract('dnssec-oracle', 'RSASHA1Algorithm')
  const SHA256Digest = loadContract('dnssec-oracle', 'SHA256Digest')
  const SHA1Digest = loadContract('dnssec-oracle', 'SHA1Digest')
  const SHA1NSEC3Digest = loadContract('dnssec-oracle', 'SHA1NSEC3Digest')

  const dnsAnchors = require('@ensdomains/dnssec-oracle/lib/anchors')
  const anchors = dnsAnchors.realEntries
  const DnsRegistrar = loadContract('dnsregistrar', 'DNSRegistrar')
  const DNSSEC = loadContract('dnssec-oracle', 'DNSSECImpl')

  /* Deploy the main contracts  */
  const dnssec = await deploy(DNSSEC, dnsAnchors.encode(anchors))
  const registrar = await deploy(DnsRegistrar, dnssec._address, ens._address)
  await ens.methods.setSubnodeOwner(ROOT_NODE, sha3('xyz'), registrar._address).send({ from: accounts[0] })

  var owner = await ens.methods.owner(namehash('xyz')).call()
  const rsasha256 = await deploy(RSASHA256Algorithm)
  const rsasha1 = await deploy(RSASHA1Algorithm)
  const sha256digest = await deploy(SHA256Digest)
  const sha1digest = await deploy(SHA1Digest)
  const sha1nsec3digest = await deploy(SHA1NSEC3Digest)
  console.log('DNSSSC ORACLE contract is deployed at ', dnssec._address)
  console.log('DNSregistrar contract is deployed at ', registrar._address)
  console.log('The owner of xyz doamin is ', owner)
  await dnssec.methods
    .setAlgorithm(5, rsasha1._address)
    .send({ from: accounts[0] })
  await dnssec.methods
    .setAlgorithm(7, rsasha1._address)
    .send({ from: accounts[0] })
  await dnssec.methods
    .setAlgorithm(8, rsasha256._address)
    .send({ from: accounts[0] })
  await dnssec.methods
    .setDigest(1, sha1digest._address)
    .send({ from: accounts[0] })
  await dnssec.methods
    .setDigest(2, sha256digest._address)
    .send({ from: accounts[0] })
  await dnssec.methods
    .setNSEC3Digest(1, sha1nsec3digest._address)
    .send({ from: accounts[0] })
  return { dnssec }
}
export default deployDNSSEC

