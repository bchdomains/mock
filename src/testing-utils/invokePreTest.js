const mockENS = require('./preTest.js').mockENS
mockENS()
