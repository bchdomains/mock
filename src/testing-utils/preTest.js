import deployENS from './deployENS.js'
import Web3 from 'web3'
import fs from 'fs'

let web3

export async function getAccounts(web3) {
  return web3.eth.getAccounts()
}

async function setupWeb3(customProvider) {
  web3 = new Web3(customProvider)
  const networkId = await web3.eth.net.getId()

  return {
    web3,
    networkId
  }
}

async function init() {
  console.log(1)
  const ENV = process.argv[2]
  const dnssec = process.argv[3] === 'dnssec'
  console.log(2)
  switch (ENV) {
    case 'GANACHE_GUI':
      var provider = new Web3.providers.HttpProvider('http://localhost:7545')
      var { web3 } = await setupWeb3(provider)
      break
    case 'GANACHE_CLI':
      var provider = new Web3.providers.HttpProvider('http://localhost:8545')
      var { web3 } = await setupWeb3(provider)
      break
    default:
      var provider = new Web3.providers.HttpProvider('http://localhost:8545')
      var { web3 } = await setupWeb3(provider)
      break
  }
  console.log(3)
  const accounts = await getAccounts(web3)
  console.log(4)
  const addresses = await deployENS({ web3, accounts, dnssec })
  console.log(5, {addresses})
  const {
    ensAddress,
    oldResolverAddresses,
    oldContentResolverAddresses,
    labels
  } = addresses
  const envLocalFile = './.env.local'
  fs.writeFileSync('./cypress.env.json', JSON.stringify(addresses))
  fs.writeFileSync(envLocalFile, `REACT_APP_ENS_ADDRESS=${ensAddress}`)
  fs.appendFileSync(envLocalFile, '\n')
  fs.appendFileSync(envLocalFile, `REACT_APP_LABELS=${JSON.stringify(labels)}`)
  fs.appendFileSync(envLocalFile, '\n')
  console.log(`Successfully wrote ENS address ${ensAddress} to .env.local`)
  fs.appendFileSync(
    envLocalFile,
    `REACT_APP_DEPRECATED_RESOLVERS=${oldResolverAddresses.join(',')}`
  )
  fs.appendFileSync(envLocalFile, '\n')
  fs.appendFileSync(
    envLocalFile,
    `REACT_APP_OLD_CONTENT_RESOLVERS=${oldContentResolverAddresses.join(',')}`
  )
  console.log(
    `Successfully wrote Old resolver address ${oldResolverAddresses.join(
      ','
    )} to .env.local`
  )
  console.log(
    `Successfully wrote Old content resolver address ${oldContentResolverAddresses.join(
      ','
    )} to .env.local`
  )
}

init()
