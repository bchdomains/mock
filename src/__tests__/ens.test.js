/**
 * @jest-environment node
 */
import GanacheCLI from './node_modules/ganache-cli'
import { setupWeb3, getAccounts } from '../web3'
import deployENS from '../testing-utils/deployENS'
import {
  getOwner,
  setOwner,
  setSubnodeOwner,
  getResolver,
  setResolver,
  getAddr,
  setAddress,
  getContent,
  setContent,
  createSubdomain,
  deleteSubdomain,
  getDomainDetails,
  getSubDomains,
  getName,
  claimAndSetReverseRecordName
} from '../registry'
import { getENS, getNamehash } from '../ens'
import '../testing-utils/extendExpect'
import Web3 from './node_modules/web3'

const ENVIRONMENTS = ['GANACHE_GUI', 'GANACHE_CLI', 'GANACHE_CLI_MANUAL']
const ENV = ENVIRONMENTS[1]

let reverseRegistrar
let baseRegistrar

describe('Blockchain tests', () => {
  beforeAll(async () => {
    switch (ENV) {
      case 'GANACHE_CLI':
        var provider = GanacheCLI.provider()
        var web3 = await setupWeb3({ provider, Web3 })
        break
      case 'GANACHE_GUI':
        var provider = new Web3.providers.HttpProvider('http://localhost:7545')
        var web3 = await setupWeb3({ provider, Web3 })
        break
      case 'GANACHE_CLI_MANUAL':
        var provider = new Web3.providers.HttpProvider('http://localhost:8545')
        var web3 = await setupWeb3({ provider, Web3 })
        break
      default:
        const options = ENVIRONMENTS.join(' or ')
        throw new Error(`ENV not set properly, please pick from ${options}`)
    }

    const accounts = await getAccounts()
    expect(accounts.length).toBeGreaterThan(0)

    const {
      ensAddress,
      reverseRegistrarAddress,
      baseRegistrarAddress
    } = await deployENS({
      web3,
      accounts
    })

    baseRegistrar = baseRegistrarAddress
    reverseRegistrar = reverseRegistrarAddress

    await getENS(ensAddress)
  }, 100000)

  describe('Test contract and Web3 setup', () => {
    test('accounts exist', async () => {
      const accounts = await getAccounts()
      expect(accounts.length).toBeGreaterThan(0)
    })

    test('ens registry, resolver and reverse registrar deployed', async () => {
      const { ENS } = await getENS()
      const accounts = await getAccounts()

      const eth = getNamehash('eth')
      const ethOwner = await ENS.owner(eth).call()
      expect(ethOwner).toBe(baseRegistrar)

      const reverse = getNamehash('reverse')
      const reverseOwner = await ENS.owner(reverse).call()
      expect(reverseOwner).toBe(accounts[0])

      const reverseNode = getNamehash('addr.reverse')
      const reverseNodeOwner = await ENS.owner(reverseNode).call()
      expect(reverseNodeOwner).toBe(reverseRegistrar)
    })
  })

  describe('Registry', () => {
    test('getOwner returns owner', async () => {
      const accounts = await getAccounts()
      const owner = await getOwner('reverse')
      expect(owner).toBe(accounts[0])
    })

    test('setSubnodeOwner sets new subnode owner', async () => {
      const owner = await getOwner('subnode.resolver.eth')
      const accounts = await getAccounts()
      expect(owner).toBe('0x0000000000000000000000000000000000000000')
      try {
        await setSubnodeOwner('subnode', 'resolver.eth', accounts[0])
        const newOwner = await getOwner('subnode.resolver.eth')
        expect(newOwner).toBe(accounts[0])
      } catch (e) {
        console.log('error', e)
      }
    })

    test('setNewOwner sets new owner', async () => {
      const owner = await getOwner('givethisaway.awesome.eth')
      const accounts = await getAccounts()
      expect(owner).toBe('0x0000000000000000000000000000000000000000')
      await setSubnodeOwner('givethisaway', 'awesome.eth', accounts[0])
      const owner2 = await getOwner('givethisaway.awesome.eth')
      expect(owner2).toBe(accounts[0])
      await setOwner('givethisaway.awesome.eth', accounts[1])
      const newOwner = await getOwner('givethisaway.awesome.eth')
      expect(newOwner).toBe(accounts[1])
    })

    test('getResolver returns a resolver address when set', async () => {
      const resolver = await getResolver('resolver.eth')
      expect(resolver).toBeHex()
      expect(resolver).toBeEthAddress()
      expect(resolver).not.toBe('0x0000000000000000000000000000000000000000')
    })

    test('getResolver returns 0x00... when resolver address is not set', async () => {
      const resolver = await getResolver('reverse')
      expect(resolver).toBeHex()
      expect(resolver).toBeEthAddress()
      expect(resolver).toBe('0x0000000000000000000000000000000000000000')
    })

    test('setResolver sets the resolver on a node', async () => {
      //test setResolver
      const resolver = await getResolver('awesome.eth')
      const mockResolver = '0x0000000000000000000000000000000000abcdef'
      expect(resolver).not.toBe(mockResolver)

      await setResolver('awesome.eth', mockResolver)
      const newResolver = await getResolver('awesome.eth')
      expect(newResolver).toBeHex()
      expect(newResolver).toBeEthAddress()
      expect(newResolver.toLowerCase()).toBe(mockResolver)
    })

    test('createSubdomain makes a new subdomain', async () => {
      const accounts = await getAccounts()
      const oldOwner = await getOwner('a.subdomain.eth')
      // expect the initial owner to be no one
      expect(oldOwner).toBe('0x0000000000000000000000000000000000000000')
      await createSubdomain('new', 'resolver.eth')
      const newOwner = await getOwner('new.resolver.eth')
      // Verify owner is the user and therefore the subdomain exists
      expect(newOwner).toBe(accounts[0])
    })

    test('deleteSubdomain deletes a subdomain', async () => {
      const accounts = await getAccounts()
      const oldOwner = await getOwner('b.subdomain.eth')
      // expect the initial owner to be no one
      expect(oldOwner).toBe('0x0000000000000000000000000000000000000000')
      await createSubdomain('b', 'subdomain.eth')
      const newOwner = await getOwner('b.subdomain.eth')
      // Verify owner is the user and therefore the subdomain exists
      expect(newOwner).toBe(accounts[0])
      await deleteSubdomain('b', 'subdomain.eth')
      const deletedOwner = await getOwner('b.subdomain.eth')
      // Verify owner has been set to 0x00... to ensure deletion
      expect(deletedOwner).toBe('0x0000000000000000000000000000000000000000')
    })
  })

  describe('Resolver', () => {
    test('getAddr returns an address', async () => {
      const addr = await getAddr('resolver.eth')
      expect(addr).toBeHex()
      expect(addr).toBeEthAddress()
      expect(addr).not.toBe('0x0000000000000000000000000000000000000000')
    })

    test('getAddr returns 0x000', async () => {
      await createSubdomain('addr', 'testing.eth')
      const resolverAddr = await getAddr('resolver.eth')
      await setResolver('addr.testing.eth', resolverAddr)
      const addr = await getAddr('addr.testing.eth')
      expect(addr).toBe('0x0000000000000000000000000000000000000000')
    })

    test('setAddr sets an address', async () => {
      //reverts if no addr is present
      const resolverAddr = await getAddr('resolver.eth')
      await setResolver('superawesome.eth', resolverAddr)
      await setAddress(
        'superawesome.eth',
        '0x0000000000000000000000000000000000012345'
      )
      const addr = await getAddr('superawesome.eth')
      expect(addr).toBe('0x0000000000000000000000000000000000012345')
    })

    // test('getContent returns a 32 byte hash', async () => {
    //   await createSubdomain('content', 'testing.eth')
    //   const content = await getContent('content.testing.eth')
    //   expect(content).toBeHex()
    //   expect(content).toMatchSnapshot()
    // })

    // test('getContent returns 0x00... when no content has been set', async () => {
    //   const content = await getContent('superawesome.eth')
    //   expect(content).toBe(
    //     '0x0000000000000000000000000000000000000000000000000000000000000000'
    //   )
    // })

    // test('setContent sets 32 byte hash', async () => {
    //   await setContent(
    //     'superawesome.eth',
    //     '0x736f6d65436f6e74656e74000000000000000000000000000000000000000000'
    //   )

    //   const content = await getContent('superawesome.eth')
    //   expect(content).toBeHex()
    //   expect(content).toMatchSnapshot()
    // })
  })

  describe('Reverse Registrar', () => {
    test('reverseNode is owned by reverseRegistrar', async () => {
      const owner = await getOwner('addr.reverse')
      expect(reverseRegistrar).toBe(owner)
    })

    test('getName gets a name for an address', async () => {
      const accounts = await getAccounts()
      const { name } = await getName(accounts[2])
      expect(name).toBe('eth')
    })

    test('claimAndSetReverseRecordName claims and sets a name', async () => {
      const accounts = await getAccounts()
      // const { name } = await getName(accounts[0])
      // expect(name).toBe(null)
      await claimAndSetReverseRecordName('resolver.eth', 2000000)
      console.log('hello')
      const { name: nameAfter } = await getName(accounts[0])
      expect(nameAfter).toBe('resolver.eth')
    })
  })

  describe('Helper functions', () => {
    test('getDomainDetails gets rootdomain and resolver details', async () => {
      const domain = await getDomainDetails('resolver.eth')
      const accounts = await getAccounts()
      expect(domain.owner).not.toBe(
        '0x0000000000000000000000000000000000000000'
      )
      expect(domain.owner).toBeEthAddress()
      expect(domain.resolver).not.toBe(
        '0x0000000000000000000000000000000000000000'
      )
      expect(domain.resolver).toBeEthAddress()
      const addr = await getAddr('resolver.eth')
      expect(domain.addr).toBe(addr)
      expect(domain.content).toMatchSnapshot()
    })

    test('getSubdomains gets all subdomains', async () => {
      const domains = await getSubDomains('eth')
      expect(domains.length).toBeGreaterThan(0)
      expect(domains[0].label).toBe('subdomain')
    })
  })
})
