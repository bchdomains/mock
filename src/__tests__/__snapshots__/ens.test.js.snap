// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain tests Helper functions getDomainDetails gets rootdomain and resolver details 1`] = `"0xe301017012204edd2984eeaf3ddf50bac238ec95c5713fb40b5e428b508fdbe55d3b9f155ffe"`;
